docs: service layer DR (#2151)
eclipse-dataspaceconnector/DataSpaceConnector: docs/developer/decision-records/2022-10-31-aggregate-service-layer/README.md

# Aggregate Service layer

## Decision

An "aggregate service" layer will be explicitly defined to provide a clear and simple interface to interact with the 
entities in a safe way.

Please note that in this document we're not referring to "services" as the components that are registered on the
runtime to achieve extensibility.

## Rationale

For `AggregateService` we intend a component that provides a clear and simple interface to interact with entities (such as
`Asset`s, `ContractNegotiation`s and so on) encapsulating some logic and features, like business validation, 
transactional context, exception handling, ...

> Currently, they are located into the `data-management-api`. Moving them to an underlying layer will permit to other kind
> of API to being exposed over them, benefiting from their offered features.

Reviewer: ```suggestion
Currently they are located into the `data-management-api`. Moving them to a separate layer will permit other kinds of APIs to use them in an equal manner benefitting from their offered features.
```

## Approach

### SPI

> The service layer will be described by a generic `aggregate-service-spi` module that, in fact will contain only the

Reviewer: ```suggestion
The service layer will be described by a generic `aggregate-service-spi` module that will contain only the
```

> `AggregateServiceResult` class, that represent a `Result` returned by an `AggregateService` method call.

Reviewer: Here, you are referring to an `AggregateService`. Is that supposed to be an actual interface? If so, I would introduce it before using it.

Author: Forgot the `*`, because `AggregateService` is not an interface, is a suffix for the aggregate service interfaces.

Every component will define its own aggregate service layer spi, for example `control-plane-spi` will contain all the
`AggregateService` interfaces of the `control-plane` as `AssetAggregateService`, `ContractNegotiationAggregateService`, 

> `TransferProcessService`, etc...

Reviewer: to keep consistency:
```suggestion
`TransferProcessAggregateService`, etc...
```

> That `control-plane-spi` module could then be used as a dependency to implement an extension that expose an API (for 
> example, a GraphQL one).

Reviewer: ```suggestion
That `control-plane-spi` module could then be used as a dependency in an extension that exposes an API (for example a GraphQL one).
```

### Implementation classes

The implementation classes will be located in a component-related module in the `core` folder, for example 

> `control-plane-aggregate-services`, this module will be referenced by the `control-plane-core` BOM to make the 
> `AggregateService`s registered at runtime.

Reviewer: ```suggestion
`control-plane-aggregate-services`, this module will be included in the `control-plane-core` BOM to make the 
`AggregateService`s available at runtime.
```

or: 

```suggestion
`control-plane-aggregate-services`, this module will be included in the `control-plane-core` BOM to register the 
`AggregateService`s.
```
